### PR TITLE
Check for data races in travis

### PR DESCRIPTION
**Pull request by gdbelvin**
This PR both adds race tests, and fixes a few outstanding race conditions. 

Blocked on #1336 -- Fixing some problem with GOPROXY

### CODE REVIEW

File: core/keyserver/keyserver.go [hunk 1]
```diff
@@ -636,36 +635,30 @@ func (s *Server) BatchQueueUserUpdate(ctx context.Context, in *pb.BatchQueueUser
...
-						errors <- status.Errorf(codes.InvalidArgument, "Invalid request")
-					}
-				}
-			}()
-		}
-	}()
-	for e := range errors {
-		return nil, e
+	g, _ := errgroup.WithContext(ctx)
+	for u := range updates {
```
Comment (pavelkalinnikov) [inline core/keyserver/keyserver.go]: Btw, the `updates` channel seems redundant here, this could iterate the `in.Updates` slice directly.

File: core/keyserver/keyserver.go [hunk 2]
```diff
@@ -636,36 +635,30 @@ func (s *Server) BatchQueueUserUpdate(ctx context.Context, in *pb.BatchQueueUser
...
 	_, tdone := monitoring.StartSpan(ctx, "BatchQueueUserUpdate.verify")
 	updates := make(chan *pb.EntryUpdate)
-	errors := make(chan error)
 	go func() {
 		defer close(updates)
 		for _, u := range in.Updates {
 			updates <- u
 		}
 	}()
-	go func() {
```
Comment (pavelkalinnikov) [inline core/keyserver/keyserver.go]: What was the race in this block? If there is none, can this be in a separate PR?

Comment (gdbelvin) [reply core/keyserver/keyserver.go]: The race in this block was that `Wait()` was being called concurrent to `Add()`

File: core/keyserver/keyserver.go (hunk 1 above)
Comment (gdbelvin) [reply core/keyserver/keyserver.go]: good point

File: core/keyserver/keyserver.go (hunk 2 above)
Comment (pavelkalinnikov) [reply core/keyserver/keyserver.go]: Doesn't look like it. All `Add`s were done synchronously, strictly before the deferred `Wait` was called.

Comment (gdbelvin) [reply core/keyserver/keyserver.go]: Sorry - I was talking about the race condition in core/client/batch_get_and_verify.go

The race condition in this block was `err = validateEntryUpdate`. The fix: `err := validateEntryUpdate`

Comment (pavelkalinnikov) [reply core/keyserver/keyserver.go]: Oh, this is a good one :)

Well the refactoring is a bit out of scope then. But LGTM, as I don't want to block you.
